src/activity/activity_heartbeat_manager.rs

```diff
@@ -265,6 +263,11 @@ impl<SG: ServerGatewayApis + Send + Sync + 'static> ActivityHeartbeatProcessor<S
                     let _ = self.cancels_tx.send(self.task_token.clone());
                 }
             }
+            // Send cancels for any activity that learns its workflow already finished (which is
+            // one thing not found implies - other reasons would seem equally valid).
+            Err(s) if s.code() == tonic::Code::NotFound => {
```

Comment:
Yeah, I had thought about attaching a reason to the cancel but didn't do it in this PR. That would make sense to add

Reply:
I have never seen a use case that requires activity performing cleanup differently if it was cancelled or the whole workflow completed. But I agree that propagating the reason for activity cancellation is still a good idea.

Reply:
Do we want to cancel in this case at all though? Or just let the activity complete?

Reply:
We talked about this in yesterday's SDK team standup.
We're going to pipe heartbeat errors back into the activity, I'll draft a design for how that's going to work.
Spencer made a good point that there's probably no point in letting an activity complete if the scheduling workflow has already been completed since there's no way to store the result.